src/buf/fixed/pool.rs

```diff
@@ -230,16 +231,17 @@ impl FixedBufPool {
 }
 
 // Internal state shared by FixedBufPool and FixedBuf handles.
-struct Inner {
+struct Inner<T> {
     // Pointer to an allocated array of iovec records referencing
     // the allocated buffers. The number of initialized records is the
     // same as the length of the states array.
     raw_bufs: ptr::NonNull<iovec>,
     // State information on the buffers. Indices in this array correspond to
     // the indices in the array at raw_bufs.
     states: Vec<BufState>,
-    // Original capacity of raw_bufs as a Vec.
-    orig_cap: usize,
+    // Original buffers, kept until drop
+    #[allow(dead_code)]
+    buffers: Vec<T>,
```

Comment:
When `T = Vec<u8>`, this adds an allocation that is completely unnecessary, since the full data of each `Vec` can be recovered from its `iovec` and `BufState`.
I'd rather not pessimize the implementations for the most common (or at least, the most convenient) parameter case.

Reply:
@mzabaluev I'm not sure about most convenient, that depends on how you allocate your memory. I've stated that for me, and probably for @FrankReh if he's using MMap's, Vec is a non starter.

However, I have realised I can address rour concern here. If I was to use `collect` rather than push, like currently, it seems collect ends up calling this specialization for vec: https://stdrs.dev/nightly/x86_64-unknown-linux-gnu/src/alloc/vec/in_place_collect.rs.html#153-213, which would avoid allocations.

Reply:
@mzabaluev I think the most recent commits address your concern for the common case

Reply:
My concern is having the extra `Vec` for keeping track of the buffers, that is redundant in the special, but important, case of `Vec` buffers. How it's populated is less important.

For carving up an mmap, I would suggest a separate collection implementing `FixedBuffers` on its own. I think `IoBufMut` is an imperfect abstraction for managing segments in a larger mapped area, since you need to keep track of the overall mmap, and adding a refcount just for that looks like impacting performance without a real need for that.

With all the concern about managing mmaps explicitly, perhaps it would be a good place to explain what are the benefits of this when applied to fixed buffers, compared to just getting buffers from the default allocator? Otherwise, this looks like a solution without a clearly stated problem.

Reply:
Ok: I specifically work pages allocated from a device driver. The _only_ way I have to access them is through a `mmap`.

Reply:
But i think you missed my point above. The extra Vec is not allocated in the special but important case of passing in a vec. It is stored, sure, but i think a point and length is not a significant overhead. In that case _no allocation for the backing memory of that Vec is performed_ the compiler should just take ownership of the impl Iterator passed in, though specialisation

Reply:
Sure, I didn't mean the buffers themselves are duplicated. But the bookkeeping in the `buffers` array is redundant in case of `Vec<u8>`, and it's still not correct: in `check_in_internal`, the buffer's initialized data length must be updated, but the current code does not do it. This "does not matter" with `Vec<u8>` buffers, but with generic `IoBufMut` I think the implementation is obliged to update it.

Reply:
I'd certianly overlooked the vec internall to FixedBuf: I think that does need changing in this PR.

I don't follow with the dual accounting though. The only location where I think the Pool or registry might need to call set_init on the IoBufMut is on the Drop of the pool or registry. My reasoning being that the buffer is wholly owned by the registry or pool whilst they are in existence, and it knows the capacity and init_len setting. If you view the Ownership -> Drop of the pool as a single operation, then its only obliged to update it before it drops, which might return the buffer elsewhere, if its a Rc backed buffer. Whilst it in the pool, nothing else has visibility of those values.

Reply:
@mzabaluev I believe I have addressed this. set_init is called before ownership of the buffers passes out of the poll or registry

Reply:
Just a comment. It's interesting this `buffers` field is used purely to be able to call set_init for each usable IoBufMut in the Inner drop.